LTO error on aarch64

When compiling / packaging cargo-udeps on Arch Linux on aarch64 I get this error. I believe this is an error on BLAKE`s part. It works without LTO.
It's not an issue on x86_64.

```
Compiling cargo-udeps v0.1.60 (/build/cargo-udeps/src/cargo-udeps-0.1.60)
error: linking with `aarch64-linux-gnu-gcc` failed: exit status: 1
  |
  = note:  "aarch64-linux-gnu-gcc" "<1 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/deps/rustcrBYtsT/libblake3-e5d4eb3908474f7e.rlib" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-ldl" "-lgit2" "-lssh2" "-lssl" "-lcrypto" "-lz" "-lcurl" "-lsqlite3" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/deps/rustcrBYtsT/raw-dylibs" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/build/blake3-273d2c9997831395/out" "-L" "/build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/build/libnghttp2-sys-06f9d6834c2e8390/out/i/lib" "-L" "<sysroot>/lib" "-L" "<sysroot>/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/deps/cargo_udeps-dd64045e507a849d" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-all" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld.lld: error: undefined symbol: blake3_hash_many_neon
          >>> referenced by cargo_udeps.1f9f9907b8f19f0f-cgu.0
          >>>               /build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/deps/cargo_udeps-dd64045e507a849d.cargo_udeps.1f9f9907b8f19f0f-cgu.0.rcgu.o:(blake3::compress_subtree_wide::hc812a1c27ae8224d)
          >>> referenced by cargo_udeps.1f9f9907b8f19f0f-cgu.0
          >>>               /build/cargo-udeps/src/cargo-udeps-0.1.60/target/release/deps/cargo_udeps-dd64045e507a849d.cargo_udeps.1f9f9907b8f19f0f-cgu.0.rcgu.o:(blake3::compress_parents_parallel::h33a4a7ed1d4d0e97)
          collect2: error: ld returned 1 exit status
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -20,10 +20,11 @@ jobs:
       fail-fast: false
       matrix:
         target: [
-          { "os": "ubuntu-latest",  "toolchain": "x86_64-unknown-linux-gnu", "name": "Linux GNU" },
-          { "os": "macOS-latest",   "toolchain": "aarch64-apple-darwin",     "name": "macOS" },
-          { "os": "windows-latest", "toolchain": "x86_64-pc-windows-msvc",   "name": "Windows MSVC" },
-          { "os": "windows-latest", "toolchain": "x86_64-pc-windows-gnu",    "name": "Windows GNU" }
+          { "os": "ubuntu-latest",    "toolchain": "x86_64-unknown-linux-gnu",  "name": "Linux GNU" },
+          { "os": "ubuntu-24.04-arm", "toolchain": "aarch64-unknown-linux-gnu", "name": "Linux GNU (arm64)" },
+          { "os": "macOS-latest",     "toolchain": "aarch64-apple-darwin",      "name": "macOS" },
+          { "os": "windows-latest",   "toolchain": "x86_64-pc-windows-msvc",    "name": "Windows MSVC" },
+          { "os": "windows-latest",   "toolchain": "x86_64-pc-windows-gnu",     "name": "Windows GNU" }
         ]
         channel: ["stable", "beta", "nightly"]
 
@@ -53,6 +54,11 @@ jobs:
     - run: cargo test --features=mmap,rayon,traits-preview,serde,zeroize
     # no_std tests.
     - run: cargo test --no-default-features
+    - name: Make sure enabling LTO via CFLAGS doesn't break the build.
+      run: cargo build --tests --features prefer_intrinsics
+      if: matrix.target.toolchain != 'x86_64-pc-windows-msvc'
+      env:
+        CFLAGS: -flto
 
     # A matrix of different test settings:
     # - debug vs release
@@ -165,10 +171,11 @@ jobs:
       fail-fast: false
       matrix:
         target: [
-          { "os": "ubuntu-latest",  "toolchain": "x86_64-unknown-linux-gnu", "name": "Linux GNU" },
-          { "os": "macOS-latest",   "toolchain": "aarch64-apple-darwin",      "name": "macOS" },
-          { "os": "windows-latest", "toolchain": "x86_64-pc-windows-msvc",   "name": "Windows MSVC" },
-          { "os": "windows-latest", "toolchain": "x86_64-pc-windows-gnu",    "name": "Windows GNU" }
+          { "os": "ubuntu-latest",    "toolchain": "x86_64-unknown-linux-gnu",  "name": "Linux GNU" },
+          { "os": "ubuntu-24.04-arm", "toolchain": "aarch64-unknown-linux-gnu", "name": "Linux GNU (arm64)" },
+          { "os": "macOS-latest",     "toolchain": "aarch64-apple-darwin",      "name": "macOS" },
+          { "os": "windows-latest",   "toolchain": "x86_64-pc-windows-msvc",    "name": "Windows MSVC" },
+          { "os": "windows-latest",   "toolchain": "x86_64-pc-windows-gnu",     "name": "Windows GNU" }
         ]
         channel: ["stable", "beta", "nightly"]
 
@@ -183,6 +190,12 @@ jobs:
       working-directory: ./b3sum
     - name: test b3sum --no-default-features
       run: cargo test --no-default-features
+      working-directory: ./b3sum
+    - name: Make sure enabling LTO via CFLAGS doesn't break the build.
+      run: cargo build --features prefer_intrinsics
+      if: matrix.target.toolchain != 'x86_64-pc-windows-msvc'
+      env:
+        CFLAGS: -flto
       working-directory: ./b3sum
 
   cross_tests:

--- a/build.rs
+++ b/build.rs
@@ -301,6 +301,20 @@ fn build_wasm32_simd() {
 }
 
 fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // If CFLAGS is set, make sure LTO is disabled by appending -fno-lto. (We can't do this with
+    // .flag() later on, because CFLAGS takes precedence.) Any mismatch between the C compiler and
+    // the Rust linker can cause linker errors in our C intrinsice implementations if LTO is
+    // enabled. See https://github.com/BLAKE3-team/BLAKE3/issues/550.
+    if !is_windows_msvc() {
+        if let Some(mut flags) = env::var_os("CFLAGS") {
+            flags.push(" -fno-lto");
+            // SAFETY: This build script is single-threaded.
+            unsafe {
+                env::set_var("CFLAGS", flags);
+            }
+        }
+    }
+
     // As of Rust 1.80, unrecognized config names are warnings. Give Cargo all of our config names.
     let all_cfgs = [
         "blake3_sse2_ffi",

